samples/cpp/train_svmsgd.cpp

```diff
@@ -0,0 +1,209 @@
+
+    currentSegment.first = Point(0, 0);
+    currentSegment.second = Point(0, height);
+    segments.push_back(currentSegment);
+}
+
+
+bool findCrossPointWithBorders(const Mat &weights, float shift, const std::pair<Point,Point> &segment, Point &crossPoint)
+{
+    int x = 0;
+    int y = 0;
+    int xMin = std::min(segment.first.x, segment.second.x);
+    int xMax = std::max(segment.first.x, segment.second.x);
+    int yMin = std::min(segment.first.y, segment.second.y);
+    int yMax = std::max(segment.first.y, segment.second.y);
+
+    CV_Assert(xMin == xMax || yMin == yMax);
+
+    if (xMin == xMax && weights.at<float>(1) != 0)
+    {
+        x = xMin;
+        y = static_cast<int>(std::floor( - (weights.at<float>(0) * x + shift) / weights.at<float>(1)));
+        if (y >= yMin && y <= yMax)
+        {
+            crossPoint.x = x;
+            crossPoint.y = y;
+            return true;
+        }
+    }
+    else if (yMin == yMax && weights.at<float>(0) != 0)
+    {
+        y = yMin;
+        x = static_cast<int>(std::floor( - (weights.at<float>(1) * y + shift) / weights.at<float>(0)));
+        if (x >= xMin && x <= xMax)
+        {
+            crossPoint.x = x;
+            crossPoint.y = y;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool findPointsForLine(const Mat &weights, float shift, Point (&points)[2], int width, int height)
+{
+    if (weights.empty())
+    {
+        return false;
+    }
+
+    int foundPointsCount = 0;
+    std::vector<std::pair<Point,Point> > segments;
+    fillSegments(segments, width, height);
+
+    for (uint i = 0; i < segments.size(); i++)
+    {
+        if (findCrossPointWithBorders(weights, shift, segments[i], points[foundPointsCount]))
+            foundPointsCount++;
+        if (foundPointsCount >= 2)
+            break;
+    }
+
+    return true;
+}
+
+void redraw(Data data, const Point points[2])
+{
+    data.img.setTo(0);
+    Point center;
+    int radius = 3;
+    Scalar color;
+    for (int i = 0; i < data.samples.rows; i++)
+    {
+        center.x = static_cast<int>(data.samples.at<float>(i,0));
+        center.y = static_cast<int>(data.samples.at<float>(i,1));
+        color = (data.responses.at<float>(i) > 0) ? Scalar(128,128,0) : Scalar(0,128,128);
+        circle(data.img, center, radius, color, 5);
+    }
+    line(data.img, points[0], points[1],cv::Scalar(1,255,1));
+
+    imshow("Train svmsgd", data.img);
+}
+
+void addPointRetrainAndRedraw(Data &data, int x, int y, int response)
+{
+    Mat currentSample(1, 2, CV_32F);
+
+    currentSample.at<float>(0,0) = (float)x;
+    currentSample.at<float>(0,1) = (float)y;
+    data.samples.push_back(currentSample);
+    data.responses.push_back(response);
+
+    Mat weights(1, 2, CV_32F);
+    float shift = 0;
+
+    if (doTrain(data.samples, data.responses, weights, shift))
+    {
+        Point points[2];
+        findPointsForLine(weights, shift, points, data.img.cols, data.img.rows);
```

Comment:
Having looked through this all, I don't think `shift` is the appropriate name. Don't you think so?